### PR TITLE
Feature/worlds

### LINKED ISSUE #37: Add support for worlds
## Summary
Add support for custom scenarios (or worlds) that will expand on usage of the bot.

## Details
Worlds are detached from each other. They have their own lorebooks, with their own characters and places and things. They would be created, maintained and chosen by slash commands.

- 1..1 relation with the `lorebook` table (so each world has its own lorebook)
- 1..1 relation with `ChannelConfig`s
- Name and description. A general, short description of the world would be sent to the AI for context
- Basic CRUD commands, as per usual
- Choice command (so the chosen world can be used in the chosen channel)
- Creation command also creates a lorebook and returns ID of both to the user
- Update command command to change world data
- Deletion command also deletes lorebook and its entries
- Make private or public

Anything else that would be important to discuss about?

### PR DESCRIPTION
**Pull request by thaalesalves**

### CODE REVIEW

File: src/main/java/es/thalesalv/chatrpg/application/mapper/world/WorldEntityToDTO.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+        return World.builder()
+                .editPermissions(t.getEditPermissions())
+                .id(t.getId())
+                .initialPrompt(t.getInitialPrompt())
+                .name(t.getName())
+                .owner(t.getOwner())
+                .visibility(t.getVisibility())
+                .description(t.getDescription())
+                .lorebook(Lorebook.builder()
```
Comment (MagicShel) [inline src/main/java/es/thalesalv/chatrpg/application/mapper/world/WorldEntityToDTO.java:36]: If there is any possibility of `t.getLorebook()` returning null, I'd use an `Optional.ofNullable(t.getLorebook()).map(l -> blah).orElse(null)` (or Collections.emptyList() if that's more appropriate).

If there is no possibility, then ignore this.

File: src/main/java/es/thalesalv/chatrpg/application/service/commands/chconfig/SetChConfigCommandService.java [hunk 2]
```diff
@@ -87,4 +93,34 @@ private ChannelEntity setChannelConfig(final String configId, final SlashCommand
...
     }
+
+    private ChannelEntity setWorld(final String worldId, final SlashCommandInteractionEvent event) {
+
+        LOGGER.debug("Attaching world ID {} to channel {} (channel ID {})", worldId,
+                event.getChannel().getName(), event.getChannel().getId());
+
+        return worldRepository.findById(worldId)
+                .map(world -> channelRepository.findByChannelId(event.getChannel().getId())
+                        .filter(channel -> null != channel)
```
Comment (MagicShel) [inline src/main/java/es/thalesalv/chatrpg/application/service/commands/chconfig/SetChConfigCommandService.java]: Replace with `.filter(Objects::nonNull)`

**Review by MagicShel — APPROVED** [review]

File: src/main/java/es/thalesalv/chatrpg/application/mapper/world/WorldEntityToDTO.java (hunk 1 above)
Comment (thaalesalves) [reply src/main/java/es/thalesalv/chatrpg/application/mapper/world/WorldEntityToDTO.java:36]: There was, in the bug I mentioned earlier for the optional worlds. The fix for the bug was to never allow it to be null buy initializing it with a dummy lorebook object that has an empty `Set<LorebookEntry>`. The nullpointers stopped, and this also allowed for entries to be added at any time while keeping worlds not required if the user doesn't want one.

Comment (thaalesalves) [reply src/main/java/es/thalesalv/chatrpg/application/mapper/world/WorldEntityToDTO.java:36]: At the initialization of these objects this happens:
```java
world.setLorebook(Optional.ofNullable(world.getLorebook())
        .map(lorebook -> setLorebook(lorebook, i.get()))
        .orElse(buildEmptyLorebook(i.get())));

// ...

    private Lorebook setLorebook(Lorebook lorebook, int id) {

        lorebook.setId(String.valueOf(id));
        lorebook.setOwner(jda.getSelfUser().getId());
        return lorebook;
    }

    private Lorebook buildEmptyLorebook(int id) {

        return Lorebook.builder()
                .id(String.valueOf(id))
                .entries(new HashSet<LorebookEntry>())
                .owner(jda.getSelfUser().getId())
                .build();
    }
```

**Pull request merged**